testcase/unittest/test_transaction_cache.py

```diff
@@ -0,0 +1,85 @@
+import os
+import random
+
+import pytest
+
+from loopchain.blockchain.exception import TransactionInvalidHashError
+from loopchain.blockchain.transactions import TransactionBuilder
+from loopchain.blockchain.transactions import TransactionVerifier
+from loopchain.blockchain.transactions import TransactionVersioner
+from loopchain.blockchain.types import ExternalAddress
+from loopchain.blockchain.types import Hash32
+from loopchain.crypto.hashing import build_hash_generator
+from loopchain.crypto.hashing import hash_generator
+from loopchain.crypto.signature import Signer
+
+tx_versioner = TransactionVersioner()
+target_attrs = ("_cache_verify_hash", "_cache_verify_signature")
+
+
+class TestTxCache:
+    @pytest.fixture
+    def tx(self):
+        """Transaction fixture"""
+        tb = TransactionBuilder.new(version="0x3", type_=None, versioner=tx_versioner )
+        tb.step_limit = 1000000
+        tb.value = 100000
+        tb.signer = Signer.from_prikey(os.urandom(32))
+        tb.to_address = ExternalAddress(os.urandom(20))
+        tb.nid = 3
+        tb.nonce = random.randint(0, 100000)
+        tb.data = "test"
+        tb.data_type = "message"
+
+        transaction = tb.build()
+
+        return transaction
+
+    def test_if_not_verified(self, tx):
+        """Check that Transaction has no attribute until verification func has been invoked"""
+        if "cache" in dir(tx):
+            raise AttributeError("Something wrong. Transaction has cached value when initialized.")
+
+        for attr in target_attrs:
+            with pytest.raises(AttributeError):
+                getattr(tx, attr)
+
+    def test_successfully_verified(self, tx):
+        """Check that verification result been cached successfully"""
+        tv = TransactionVerifier.new("0x3",  tx.type(), tx_versioner)
+
+        tv.verify_hash(tx)
+        tv.verify_signature(tx)
+
+        for attr in target_attrs:
+            assert getattr(tx, attr)
+
+    def test_raised_exception_when_verified(self, tx, monkeypatch):
+        """Check that exception successfully cached when raised exception in verification func"""
+        fake_hash: Hash32 = Hash32.new()
+        origin_generate_hash = hash_generator.HashGenerator.generate_hash
+        def monkey_generate_hash(self, origin_data: dict):
+            return fake_hash
+
+        # Monkeypatch
+        monkeypatch.setattr(hash_generator.HashGenerator, "generate_hash", monkey_generate_hash)
```

Comment:
This is too complicated.
I suggest changing `hash` attribute of `Transaction`

Reply:
Tried to fix it. Please check.